examples/ctc.dx

```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
     Left () -> first
     Right i -> seq.i
 
-def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & (Fin 2)))=>v) =
-  -- Turns "text" into " t e x t".
-  -- The output of this function has a slightly complicated output type, which
-  -- has size 1 + 2 * (size m).
-  interleaved = interleave blank seq
-  prepend blank interleaved
-
-def clipidx (n:Type) [Ix n] (i:Int) : n =
-  -- Returns element at 0 if less than zero.
-  -- Ideally we could have an alternative
-  -- to Fin that just clips the index at its bounds.
-  from_ordinal n $ unsafe_i_to_n (select (i < 0) 0 i)
-
-def logaddexp (x:Float) (y:Float) : Float =
-  m = max x y
-  m + ( log ( (exp (x - m) + exp (y - m))))
-
-def ctc {vocab time position} [Eq vocab, Eq position, Eq time]
-      (blank:  vocab)
-      (logits: time=>vocab=>Float)
-      (labels: position=>vocab)
-      : Float =
-  -- Computes log p(labels | logits), marginalizing over possible alignments.
-  -- Todo: remove unnecessary implicit type annotations once
-  -- Dex starts putting implicit types in scope.
+def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & IsBlank))=>v) =
+  -- Turns "text" into " t e x t ".
+  prepend blank (interleave blank seq)
 
-  ilabels = prepend_and_interleave blank labels
-
-  normalized_logits = for t. logsoftmax logits.t
-
-  -- Initialization rules
-  logprob_start_with_blank = normalized_logits.(0@_).blank
-  logprob_of_first_label   = normalized_logits.(0@_).(labels.(0@_))
-  log_prob_seq_t0 = for pos.
-    -- TODO: allow pattern-matching on integer literals
-    case ordinal pos == 0 of
-      True -> logprob_start_with_blank
-      False -> case ordinal pos == 1 of
-        True  -> logprob_of_first_label
-        False -> log 0.000001
-
-  same_as_last = \ilabels s.
-    o = n_to_i $ ordinal s
-    select (o >= 2) (ilabels.s == ilabels.(clipidx _ (o - 2))) False
-
-  safe_idx = \prev s.
-    select (s >= 0) prev.(clipidx _ s) (log 0.0)
-
-  -- Todo: As suggested by Adam Paske, we could get rid of these
-  -- logaddexp calls with a newtype that overloads + and *
-  update = \t prev.
-    case ordinal t == 0 of
-      True -> log_prob_seq_t0
-      False -> for s.
-        cond = ilabels.s == blank || same_as_last ilabels s
-        labar = logaddexp prev.s (safe_idx prev (n_to_i (ordinal s) - 1))
-        other = logaddexp labar  (safe_idx prev (n_to_i (ordinal s) - 2))
-        ans = select cond labar other
-        ans + normalized_logits.t.(ilabels.s)
+def normalize {n v} [Fractional v, Add v] (xs:n=>v) : n=>v =
+  for i. divide xs.i (sum xs)
```

Comment:
Dex is not yet smart enough yet to lift this `sum xs` out of the enclosing `for` on its own, so as written this does quadratic work.  Recommend computing the sum outside the `for`.

Reply:
Ah ok, will fix.  Although I guess I view my role as example-writer is partly to write the code that I wish would be fast, in order to spur on compiler development :)